software/nosql/src/main/java/org/apache/brooklyn/entity/nosql/mongodb/AbstractMongoDBSshDriver.java

```diff
@@ -73,14 +75,40 @@ public void install() {
     public void customize() {
         Map<?,?> ports = ImmutableMap.of("port", getServerPort());
         Networking.checkPortsValid(ports);
-        String command = String.format("mkdir -p %s", getDataDirectory());
+        List<String> commands = new LinkedList<>();
+        commands.add(String.format("mkdir -p %s", getDataDirectory()));
+
+        if (MongoDBAuthenticationUtils.usesAuthentication(entity)) {
+            String destinationLocation = Os.mergePaths(getRunDir(), "mongodb-keyfile");
+            entity.sensors().set(AbstractMongoDBServer.MONGODB_KEYFILE_DESTINATION, destinationLocation);
+            String keyfileContents = entity.config().get(AbstractMongoDBServer.MONGODB_KEYFILE_CONTENTS);
+            if (keyfileContents == null) {
+                String keyfileUrl = entity.config().get(AbstractMongoDBServer.MONGODB_KEYFILE_URL);
+                copyResource(keyfileUrl, destinationLocation);
```

Comment:
I think it's worth checking that `keyfileUrl` isn't null too just in case someone changes the behaviour of `usesAuthentication` later on without realising that this block implicitly relies on `keyfileUrl` being non-null.